File: android/shared/src/main/java/dev/scorpion7slayer/modelsmeter/UsageHistory.java
package dev.scorpion7slayer.modelsmeter;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/** Bounded, local-only usage samples used by charts and trend-aware estimates. */
public final class UsageHistory {
    public static final String FIVE_HOUR = "five_hour";
    public static final String WEEKLY = "weekly";
    public static final String MONTHLY = "monthly";
    private static final int MAX_FIVE_HOUR_SAMPLES = 288;
    private static final int MAX_WEEKLY_SAMPLES = 336;
    private static final int MAX_MONTHLY_SAMPLES = 372;

    public final String kind;
    public final List<UsageSample> samples;

    public UsageHistory(String kind, List<UsageSample> samples) {
        this.kind = normalizeKind(kind);
        this.samples = Collections.unmodifiableList(new ArrayList<>(
                samples == null ? Collections.emptyList() : samples));
    }

    public static UsageHistory empty(String kind) {
        return new UsageHistory(kind, Collections.emptyList());
    }

    public UsageHistory append(UsageWindow window, long observedAtMillis) {
        if (window == null || observedAtMillis <= 0L || window.windowSeconds <= 0L) return this;
        long resetAt = window.effectiveResetAtMillis(observedAtMillis);
        if (resetAt <= observedAtMillis) return this;
        UsageSample next = new UsageSample(observedAtMillis, window.usedPercent, resetAt,
                window.windowSeconds);
        ArrayList<UsageSample> updated = new ArrayList<>(samples);
        if (!updated.isEmpty()) {
            UsageSample last = updated.get(updated.size() - 1);
            if (observedAtMillis <= last.observedAtMillis) return this;
            long minimumSpacing = FIVE_HOUR.equals(kind)
                    ? TimeUnit.MINUTES.toMillis(5)
                    : MONTHLY.equals(kind)
                            ? TimeUnit.HOURS.toMillis(2) : TimeUnit.MINUTES.toMillis(30);
            if (sameWindow(last, next) && last.usedPercent == next.usedPercent
                    && observedAtMillis - last.observedAtMillis < minimumSpacing) {
                updated.set(updated.size() - 1, next);
                return new UsageHistory(kind, updated);
            }
        }
        updated.add(next);
        int max = maximumSamples(kind);
        while (updated.size() > max) updated.remove(0);
        return new UsageHistory(kind, updated);
    }

    /** Samples belonging to the latest reset window, ordered oldest to newest. */
    public List<UsageSample> currentWindowSamples() {
        if (samples.isEmpty()) return Collections.emptyList();
        UsageSample latest = samples.get(samples.size() - 1);
        ArrayList<UsageSample> result = new ArrayList<>();
        for (UsageSample sample : samples) {
            if (sameWindow(sample, latest)) result.add(sample);
        }
        return Collections.unmodifiableList(result);
    }

    /** Most recent reset windows, oldest to newest, for normalized historical overlays. */
    public List<List<UsageSample>> recentWindows(int maximum) {
        if (samples.isEmpty() || maximum <= 0) return Collections.emptyList();
        ArrayList<List<UsageSample>> windows = new ArrayList<>();
        ArrayList<UsageSample> current = new ArrayList<>();
        UsageSample previous = null;
        for (UsageSample sample : samples) {
            if (previous != null && !sameWindow(previous, sample)) {
                windows.add(Collections.unmodifiableList(current));
                current = new ArrayList<>();
            }
            current.add(sample);
            previous = sample;
        }
        if (!current.isEmpty()) windows.add(Collections.unmodifiableList(current));
        int from = Math.max(0, windows.size() - maximum);
        return Collections.unmodifiableList(new ArrayList<>(windows.subList(from, windows.size())));
    }

    public int completedWindowCount() {
        if (samples.isEmpty()) return 0;
        int boundaries = 0;
        UsageSample previous = samples.get(0);
        for (int i = 1; i < samples.size(); i++) {
            UsageSample next = samples.get(i);
            if (!sameWindow(previous, next)) boundaries++;
            previous = next;
        }
        return boundaries;
    }

    /**
     * Returns a recent observed burn rate in percent per millisecond. Samples must span enough
     * time and percentage points to avoid magnifying refresh jitter.
     */
    public double observedBurnRate() {
        List<UsageSample> current = currentWindowSamples();
        if (current.size() < 2) return 0d;
        UsageSample latest = current.get(current.size() - 1);
        long minimumSpan = FIVE_HOUR.equals(kind)
                ? TimeUnit.MINUTES.toMillis(10)
                : MONTHLY.equals(kind) ? TimeUnit.HOURS.toMillis(6) : TimeUnit.HOURS.toMillis(2);
        UsageSample first = null;
        for (UsageSample candidate : current) {
            if (latest.observedAtMillis - candidate.observedAtMillis >= minimumSpan
                    && latest.usedPercent - candidate.usedPercent >= 2) {
                first = candidate;
                break;
            }
        }
        if (first == null) return 0d;
        return (latest.usedPercent - first.usedPercent)
                / (double) (latest.observedAtMillis - first.observedAtMillis);
    }

    public JSONObject toJson() throws JSONException {
        JSONArray array = new JSONArray();
        for (UsageSample sample : samples) array.put(sample.toJson());
        return new JSONObject().put("version", 1).put("kind", kind).put("samples", array);
    }

    public static UsageHistory fromJson(JSONObject json, String fallbackKind) {
        if (json == null) return empty(fallbackKind);
        String kind = normalizeKind(json.optString("kind", fallbackKind));
        JSONArray array = json.optJSONArray("samples");
        ArrayList<UsageSample> samples = new ArrayList<>();
        if (array != null) {
            for (int i = 0; i < array.length(); i++) {
                UsageSample sample = UsageSample.fromJson(array.optJSONObject(i));
                if (sample != null) samples.add(sample);
            }
        }
        int max = maximumSamples(kind);
        if (samples.size() > max) {
            samples = new ArrayList<>(samples.subList(samples.size() - max, samples.size()));
        }
        return new UsageHistory(kind, samples);
    }

    private static int maximumSamples(String kind) {
        if (FIVE_HOUR.equals(kind)) return MAX_FIVE_HOUR_SAMPLES;
        if (MONTHLY.equals(kind)) return MAX_MONTHLY_SAMPLES;
        return MAX_WEEKLY_SAMPLES;
    }

    static boolean sameWindow(UsageSample left, UsageSample right) {
        if (left == null || right == null) return false;
        return UsageWindow.sameResetWindow(left.resetAtMillis, left.windowSeconds,
                right.resetAtMillis, right.windowSeconds);
    }

    private static String normalizeKind(String kind) {
        if (WEEKLY.equals(kind)) return WEEKLY;
        if (MONTHLY.equals(kind)) return MONTHLY;
        return FIVE_HOUR;
    }
}

File: ios/ModelsMeterCore/Sources/ModelsMeterCore/UsageHistory.swift
import Foundation

public enum UsageHistoryKind: String, Codable, Sendable, CaseIterable {
    case fiveHour = "five_hour"
    case weekly
    case monthly
}

/// One locally observed allowance value. It contains no account or credential data.
public struct UsageSample: Codable, Sendable, Equatable, Identifiable {
    public var id: Date { observedAt }

    public let observedAt: Date
    public let usedPercent: Int
    public let resetAt: Date
    public let windowSeconds: Int64

    public init(
        observedAt: Date,
        usedPercent: Int,
        resetAt: Date,
        windowSeconds: Int64
    ) {
        self.observedAt = observedAt
        self.usedPercent = min(100, max(0, usedPercent))
        self.resetAt = resetAt
        self.windowSeconds = max(0, windowSeconds)
    }

    private enum CodingKeys: String, CodingKey {
        case observedAt
        case usedPercent
        case resetAt
        case windowSeconds
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.init(
            observedAt: try container.decodeIfPresent(Date.self, forKey: .observedAt)
                ?? Date(timeIntervalSince1970: 0),
            usedPercent: try container.decodeIfPresent(Int.self, forKey: .usedPercent) ?? 0,
            resetAt: try container.decodeIfPresent(Date.self, forKey: .resetAt)
                ?? Date(timeIntervalSince1970: 0),
            windowSeconds: try container.decodeIfPresent(Int64.self, forKey: .windowSeconds) ?? 0
        )
    }

    fileprivate var isValid: Bool {
        observedAt.timeIntervalSince1970.isFinite
            && observedAt.timeIntervalSince1970 > 0
            && resetAt.timeIntervalSince1970.isFinite
            && resetAt > observedAt
            && windowSeconds > 0
    }
}

/// Bounded, local-only samples used by usage charts and trend-aware pace estimates.
public struct UsageHistory: Codable, Sendable, Equatable {
    public let kind: UsageHistoryKind
    public let samples: [UsageSample]

    public init(kind: UsageHistoryKind, samples: [UsageSample]) {
        self.kind = kind
        let maximum = Self.maximumSamples(for: kind)
        let normalized = samples
            .filter(\.isValid)
            .sorted { $0.observedAt < $1.observedAt }
        self.samples = Array(normalized.suffix(maximum))
    }

    public static func empty(_ kind: UsageHistoryKind) -> Self {
        Self(kind: kind, samples: [])
    }

    public func appending(window: UsageWindow, observedAt: Date) -> Self {
        guard observedAt.timeIntervalSince1970.isFinite,
              observedAt.timeIntervalSince1970 > 0,
              window.windowSeconds > 0,
              let resetAt = window.effectiveResetDate(relativeTo: observedAt),
              resetAt > observedAt else {
            return self
        }

        let next = UsageSample(
            observedAt: observedAt,
            usedPercent: window.usedPercent,
            resetAt: resetAt,
            windowSeconds: window.windowSeconds
        )
        var updated = samples
        if let last = updated.last {
            guard observedAt > last.observedAt else {
                return self
            }
            let minimumSpacing: TimeInterval = switch kind {
            case .fiveHour: 5 * 60
            case .weekly: 30 * 60
            case .monthly: 2 * 60 * 60
            }
            if Self.sameWindow(last, next),
               last.usedPercent == next.usedPercent,
               observedAt.timeIntervalSince(last.observedAt) < minimumSpacing {
                updated[updated.count - 1] = next
                return Self(kind: kind, samples: updated)
            }
        }

        updated.append(next)
        return Self(kind: kind, samples: updated)
    }

    /// Samples from the latest reset window, oldest first.
    public var currentWindowSamples: [UsageSample] {
        guard let latest = samples.last else { return [] }
        return samples.filter { Self.sameWindow($0, latest) }
    }

    /// The most recent reset windows, oldest window first and oldest sample first.
    public func recentWindows(maximum: Int) -> [[UsageSample]] {
        guard maximum > 0, !samples.isEmpty else { return [] }
        var windows: [[UsageSample]] = []
        var current: [UsageSample] = []
        var previous: UsageSample?

        for sample in samples {
            if let previous, !Self.sameWindow(previous, sample) {
                windows.append(current)
                current = []
            }
            current.append(sample)
            previous = sample
        }
        if !current.isEmpty {
            windows.append(current)
        }
        return Array(windows.suffix(maximum))
    }

    public var completedWindowCount: Int {
        guard var previous = samples.first else { return 0 }
        var count = 0
        for sample in samples.dropFirst() {
            if !Self.sameWindow(previous, sample) {
                count += 1
            }
            previous = sample
        }
        return count
    }

    /// Recent observed burn in percentage points per second.
    public var observedBurnRate: Double {
        let current = currentWindowSamples
        guard let latest = current.last, current.count >= 2 else { return 0 }
        let minimumSpan: TimeInterval = switch kind {
        case .fiveHour: 10 * 60
        case .weekly: 2 * 60 * 60
        case .monthly: 6 * 60 * 60
        }
        guard let first = current.first(where: {
            latest.observedAt.timeIntervalSince($0.observedAt) >= minimumSpan
                && latest.usedPercent - $0.usedPercent >= 2
        }) else {
            return 0
        }
        let elapsed = latest.observedAt.timeIntervalSince(first.observedAt)
        guard elapsed > 0 else { return 0 }
        return Double(latest.usedPercent - first.usedPercent) / elapsed
    }

    private enum CodingKeys: String, CodingKey {
        case kind
        case samples
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.init(
            kind: try container.decodeIfPresent(UsageHistoryKind.self, forKey: .kind) ?? .fiveHour,
            samples: try container.decodeIfPresent([UsageSample].self, forKey: .samples) ?? []
        )
    }

    private static func maximumSamples(for kind: UsageHistoryKind) -> Int {
        switch kind {
        case .fiveHour: 288
        case .weekly: 336
        case .monthly: 372
        }
    }

    private static func sameWindow(_ left: UsageSample, _ right: UsageSample) -> Bool {
        UsageWindow.sameResetWindow(
            leftReset: left.resetAt,
            leftWindowSeconds: left.windowSeconds,
            rightReset: right.resetAt,
            rightWindowSeconds: right.windowSeconds
        )
    }
}
